SuccessBonusActivationActivity/handler.ts

```diff
@@ -0,0 +1,181 @@
+  "SuccessBonusActivationFailure"
+);
+export type SuccessBonusActivationFailure = t.TypeOf<
+  typeof SuccessBonusActivationFailure
+>;
+
+export const SuccessBonusActivationResult = t.taggedUnion("kind", [
+  SuccessBonusActivationSuccess,
+  SuccessBonusActivationFailure
+]);
+export type SuccessBonusActivationResult = t.TypeOf<
+  typeof SuccessBonusActivationResult
+>;
+
+/**
+ * Converts a Promise<Either> into a TaskEither
+ * This is needed because our models return unconvenient type. Both left and rejection cases are handled as a TaskEither left
+ * @param lazyPromise a lazy promise to convert
+ *
+ * @returns either the query result or a query failure
+ */
+const fromQueryEither = <R>(
+  lazyPromise: () => Promise<Either<QueryError | Error, R>>
+): TaskEither<Error, R> =>
+  tryCatch(lazyPromise, toError).chain(errorOrResult =>
+    fromEither(errorOrResult).mapLeft(toError)
+  );
+
+const updateBonusAsActive = (
+  bonusActivationModel: BonusActivationModel,
+  bonusActivation: BonusActivationWithFamilyUID
+): TaskEither<Error, RetrievedBonusActivation> => {
+  return fromQueryEither(() => {
+    const bonusToUpdate: BonusActivationWithFamilyUID = {
+      ...bonusActivation,
+      status: BonusActivationStatusEnum.ACTIVE
+    };
+    return bonusActivationModel.replace(bonusToUpdate);
+  });
+};
+
+const saveBonusForEachFamilyMember = (
+  userBonusModel: UserBonusModel,
+  {
+    applicantFiscalCode,
+    id: bonusId,
+    dsuRequest: { familyMembers }
+  }: BonusActivationWithFamilyUID
+): TaskEither<Error, readonly RetrievedUserBonus[]> =>
+  array.sequence(taskEither)(
+    familyMembers
+      .map<NewUserBonus>(({ fiscalCode }) => ({
+        bonusId,
+        fiscalCode,
+        id: (bonusId as unknown) as NonEmptyString,
+        isApplicant: applicantFiscalCode === fiscalCode,
+        kind: "INewUserBonus"
+      }))
+      .map(nub => () => userBonusModel.create(nub, bonusId))
+      .map(fn => fromQueryEither(fn))
+  );
+
+type ISuccessBonusActivationHandler = (
+  context: Context,
+  input: unknown
+) => Promise<SuccessBonusActivationResult>;
+/**
+ * Operations to be perfomed in case the bonus request was accepted by ADE
+ *
+ *
+ * @returns either a success or a failure request
+ * @throws when the response is considered a transient failure and thus is not considered a domain message
+ */
+export function SuccessBonusActivationHandler(
+  bonusActivationModel: BonusActivationModel,
+  userBonusModel: UserBonusModel
+): ISuccessBonusActivationHandler {
+  return async (
+    context: Context,
+    input: unknown
+  ): Promise<SuccessBonusActivationResult> => {
+    return taskEither
+      .of<SuccessBonusActivationFailure, void>(void 0)
+      .chain(_ =>
+        fromEither(SuccessBonusActivationInput.decode(input))
+          .mapLeft(() => InvalidInputFailure.encode({ kind: "INVALID_INPUT" }))
+          .map(({ bonusActivation }) => bonusActivation)
+      )
```

Comment:
@balanza what do you think about this?
```suggestion
    return fromEither(SuccessBonusActivationInput.decode(input))
      .mapLeft<SuccessBonusActivationFailure>(() =>
        InvalidInputFailure.encode({ kind: "INVALID_INPUT" })
      )
      .map(({ bonusActivation }) => bonusActivation)
```

Reply:
The reason why I use `taskEither.of<SuccessBonusActivationFailure, void>(void 0)` is that this way I have a left-type which is extended by every further step in the chain. Otherwise, at every step I'd need to lift the left type (with `foldTaskEither` usually).
Did you try this code: I'll do, but I think it won't compile

Reply:
It compiles. The typing into the `.mapLeft<SuccessBonusActivationFailure>` change the left type for all the chain as `taskEither.of<SuccessBonusActivationFailure, void>(void 0)` do

Reply:
ok, I was wrong: it does compile. I'll merge your suggestion, thanks